README.md
# dzcb

**D**MR **Z**one **C**hannel **B**uilder

* Fetch - fetch input files from local directories or urls
* Assemble - combine information from multiple sources
* Filter - rename, exclude, reorder zones, and talkgroups
* Format - output to common export formats

## [Overview Video](https://youtu.be/RfokJM5rpsM)

<img src="/doc/dzcb-overview.svg">

## See [Releases](https://github.com/mycodeplug/dzcb/releases) for Default Codeplugs

## Github Actions

For more information on generating customized codeplugs in the cloud without
installing any software, see **[WALKTHROUGH](./doc/WALKTHROUGH.md)**.

The walkthrough uses
[mycodeplug/example-codeplug](https://github.com/mycodeplug/example-codeplug)
[[input files](https://github.com/mycodeplug/example-codeplug/tree/main/input/default)]

# Output Formats

## Farnsworth JSON

For import into [editcp](https://www.farnsworth.org/dale/codeplug/editcp/),
with theoretical support for:

* TYT
  * MD-380, MD-390
  * MD-UV380 (tested), MD-UV390
  * MD-2017 
* Alinco DJ-MD40 
* Retevis RT3, RT3-G, RT3S, and RT82 radios

Use `--farnsworth-template-json` to specify an exported codeplug to
use as the template for settings and radio capabilities.

`dzcb` includes basic template files for [MD380/390](./codeplug/default-tyt-md380)
and [MD-UV380/390](./src/dzcb/data/farnsworth).

Generated JSON files will be written to the `editcp` subdir of the output directory.

## [OpenRTX/dmrconfig](https://github.com/OpenRTX/dmrconfig)

For import into [`dmrconfig`](https://github.com/OpenRTX/dmrconfig) 1.1
utility with theoretical support for:

* Anytone AT-D868UV
* Anytone AT-D878UV
* BTECH DMR-6x2
* Baofeng DM-1801
* Radioddity GD-77
* TYT MD-380
* TYT MD-390
* Zastone D900
* Zastone DP880
* Radtel RT-27D
* Baofeng RD-5R
* TYT MD-UV380
* TYT MD-UV390
* TYT MD-2017
* TYT MD-9600
* Retevis RT84

Use `--dmrconfig-template` to specify an exported codeplug config to use as a
template for radio type, messages, DMR ID,  and startup text. The `dmrconfig` subdir
of the output directory will contain a dmrconfig config per template specified.

## Anytone CPS

For import into the official Anytone CPS (windows-only).

`dzcb` generates Channels.CSV, Talkgroups.CSV, ScanList.CSV, and Zone.CSV in the `anytone` subdir
of the output directory.

### Versions Supported

* **AT-D578UV** [CPS 1.11](https://cdn.shopify.com/s/files/1/0833/9095/files/D578UV_V1.11_official_release_200918.zip)
  * [Bridgecom Support](https://bridgecomsystems.freshdesk.com/support/solutions/articles/63000106309-anytone-578-cps-and-firmware-downloads)
* **AT-D868UV** [CPS 1.39](https://cdn.shopify.com/s/files/1/0833/9095/files/D868UV_2.39_official_200807.zip)
  * [Bridgecom Support](https://bridgecomsystems.freshdesk.com/support/solutions/articles/63000105671-anytone-868-cps-firmware-and-driver-versions)
* **AT-D878UV** [CPS 1.21](https://cdn.shopify.com/s/files/1/0833/9095/files/D878UV_V1.21_official_release_200918.zip)  
  * [Bridgecom Support](https://bridgecomsystems.freshdesk.com/support/solutions/articles/63000105978-anytone-878-878-plus-cps-firmware-and-driver-downloads)

(links fetched 2021/01/24)

## GB3GF OpenGD77 CSV

For import into [GB3GF CSV tool](http://www.gb3gf.co.uk/downloads.html).
Currently only supporting OpenGD77 target. Tool is "windows-only" but
runs decently under wine.

If `--gb3gf` is specified, the `gb3gf/opengd77` subdir of the output directory will
contain the 4 CSV files used by the program:

  * `Channels.csv`
  * `Contacts.csv`
  * `TG_Lists.csv`
  * `Zones.csv`
  
Note: these files are actually semicolon separated.

# Input Format

## K7ABD style

A directory of inter-related CSV files describing the common settings
needed to program all radios. This is a common format used by K7ABD's
original [anytone-config-builder](https://www.k7abd.net/anytone-config-builder/)
and N7EKB's [cps-import-builder](https://github.com/n7ekb/cps-import-builder).

An arbitrary number of files and folders will be combined. At this time duplicate
zone names in different files may be problematic. Keep zone names unique.
Particularly be wary of trucated names.

* `Analog__ZoneName.csv`
  * Zone, Channel Name, Bandwidth, Power, RX Freq, TX Freq, CTCSS Decode, CTCSS Encode, TX Prohibit
* `Talkgroups__ZoneName.csv`
  * **No Header Row** tuples of: talkgroup_name,talkgroup_number
  * To specify a **Private** call, suffix the talkgroup number with a "P"
* `Digital-Other__ZoneName.csv` -- only really useful for digital simplex channels or private call
  * Zone, Channel Name, Power, RX Freq, TX Freq, Color Code, Talk Group, TimeSlot, Call Type, TX Permit
* `Digital-Repeaters__ZoneName.csv` -- each line is a digital channel with static talkgroup timeslot assignments
  * Zone Name, Comment, Power, RX Freq, TX Freq, Color Code, talkgroup1, talkgroup2, talkgroup3, ...
  * Typically the Zone Name field is semicolon separated: "Longer Name;LNM"
  * The value for the talkgroup column should be "-", "1", or "2"
  * The talkgroup names must exist in the talkgroups file.
  * Some codeplug targets will create a zone for each frequency with channels for each static talkgroup.
  
## Data Sources

### [PNWdigital.net](http://PNWDigital.net)

**Before using this network, please read the [quick start](http://www.pnwdigital.net/quick-start.html)**

When building a codeplug with the `--pnwdigital` switch, 
data is downloaded from [PNWDigital.net/files/acb](http://www.pnwdigital.net/files/acb/).
The files are updated regularly from the cBridge, so there
could be test data or repeaters that are not yet active.

The data is cached by [`dzcb.pnwdigital`](./src/dzcb/pnwdigital.py)

### [SeattleDMR](https://seattledmr.org/)

**Before using these repeaters, please read the website**

When building a codeplug with the `--seattledmr` switch,
data is downloaded from [seattledmr.org/ConfigBuilder/Digital-Repeaters-Seattle-addon.csv](https://seattledmr.org/ConfigBuilder/Digital-Repeaters-Seattle-addon.csv)
and cleaned up a bit in [`dzcb.seattledmr`](./src/dzcb/seattledmr.py)

### Repeaterbook Proximity

Download live analog Repeaterbook data within distance of point of
interest.

`--repeaterbook-proximity-csv` references a csv file with the fields:

* Zone Name,Lat,Long,Distance,Unit,Band(2m;1.25m;70cm),Use,Operational Status,etc

See example: [`src/dzcb/data/repeaterbook_proximity_zones.csv`](/src/dzcb/data/repeaterbook_proximity_zones.csv)

The fields after Band are optional and correspond directly to the field names
and values in the [Repeaterbook API](https://www.repeaterbook.com/wiki/doku.php?id=api)
(see examples).

`--repeaterbook-state` is a space-separated list of US states or Canadian
provinces that should be included in the proximity search. Including more
states will increase the time required to generate the codeplug.

Repeaterbook API data is downloaded and cached in a user and platform-specific
cache directory. Data will be refreshed if it is older than 12 hours. When
downloading from Repeaterbook, a delay of 30 seconds is introduced between
requests to reduce load on the repeaterbook servers.

Please respect their servers and submit changes requests to repeaterbook
directly.

See [`dzcb.repeaterbook`](./src/dzcb/repeaterbook.py).

#### Example Format

```
Zone Name,Lat,Long,Distance,Unit,Band(2m;1.25m;70cm),Use,Operational Status
Longview WA 35mi,46.13819885,-122.93800354,35,miles,2m;70cm,OPEN,On-air
Longview WA VHF 35mi,46.13819885,-122.93800354,35,miles,2m,open,On-air
Longview WA UHF 35mi,46.13819885,-122.93800354,35,miles,70cm,OPEN,On-Air
```

(it's easy to search on repeaterbook and copy the info from the URL!)

### Simplex, GMRS, etc

Some common [Digital](./src/dzcb/data/k7abd/Digital-Others__Simplex.csv)
and [Analog](./src/dzcb/data/k7abd/Analog__Simplex.csv) simplex frequencies,
and [GMRS/FRS and MURS channels](./src/dzcb/data/k7abd/Analog__Unlicensed.csv) are
included if `--default-k7abd` is specified.

# Customization

The channels, zones, and contacts present and the ordering of such in the
final generated codeplug is controlled by one or more CSV files passed
as parameters.

The column headers correspond to the codeplug objects (zones, contacts, channels,
etc) and may appear in any order or be omitted (if no change is requested to
that type of codeplug object).

The names used in include, exclude, and order are *python regular expressions*
that match from the beginning of the object name. They are prefix match by
default.  To opt-out of prefix match, add a `$` at the end. All regular
expression special characters should be escaped (`\`).

## `--include`

If the CSV file is passed as a `--include` parameter, then any objects not explicitly
mentioned in the file will be removed.

## `--exclude`

If the CSV file is passed as an `--exclude` parameter, then any objects mentioned
will be removed.

See example: [`codeplug/default-tyt-md380/exclude.csv`](/codeplug/default-tyt-md380/exclude.csv)
(for radios with lower channel/zone capacity)

## `--order`

Alphabetic sorting by filename is used when initially building the zones and channels
from K7ABD input files. No implicit sorting by object name occurs.

Any remaining objects are sorted according to the order specified in the file
passed to `--order`. Similarly, a file may be specified as `--reverse-order` which
will sort in reverse, starting from the end of the list.

See example: [`codeplug/default/order.csv`](/codeplug/default/order.csv)

## `--replacements`

The replacements CSV column headers are like `object_pattern,object_repl` and
correspond exactly to the arguments passed to python's
[`re.sub`](https://docs.python.org/3/library/re.html#re.sub). Similar to the
ordering csv, `object` would refer to zones, contacts, channels, etc.

The replacements will be applied to the name of each object of the given
type in the listed order. Subsequent patterns will affect previous
replacements. Unlike the previous group, these regular expressions are
case sensitive, support capturing groups, and indexed replacement.

See example: [`codeplug/default/replacements.csv`](/codeplug/default/replacements.csv)

## `--repeaterbook-name-format`

Python format string used to generate channel names from repeaterbook.

The format string will be passed the Repeaterbook API response dictionary
for the repeater, so all valid fields are usable in the format string.

The default name format is: `"{Callsign} {Nearest City} {Landmark}"`

# Bonus: Trim the Contact List

Download the usersDB.bin with Farnsworth editcp.

    python -m dzcb.contact_trim \
        < "~/.cache/codeplug/Codeplug Editor/usersDB.bin" \
        > "~/.cache/codeplug/Codeplug Editor/usersDB-trimmed.bin"


# Basic Usage

## Installation

```
pip install dzcb
```

## Example

Download live pnwdigital and seattledmr networks and generate a dmrconfig style
codeplug in `/tmp/my-codeplug` with local simplex zones included.

```
python -m dzcb \
    --pnwdigital \
    --seattledmr \
    --default-k7abd \
    --dmrconfig -- /tmp/my-codeplug
```

# Complex Usage

See **[WALKTHROUGH](./doc/WALKTHROUGH.md)** for a step-by-step guide
to getting started with your own customized codeplug build. (The walkthrough
uses [mycodeplug/example-codeplug](https://github.com/mycodeplug/example-codeplug))

## CLI

```
$ python -m dzcb --help
usage: python -m dzcb [-h] [--pnwdigital] [--seattledmr] [--default-k7abd] [--k7abd [DIR [DIR ...]]]
                   [--repeaterbook-proximity-csv [CSV [CSV ...]]] [--repeaterbook-state [STATE [STATE ...]]]
                   [--repeaterbook-name-format REPEATERBOOK_NAME_FORMAT] [--scanlists-json JSON] [--include [CSV [CSV ...]]]
                   [--exclude [CSV [CSV ...]]] [--order [CSV [CSV ...]]] [--reverse-order [CSV [CSV ...]]]
                   [--replacements [CSV [CSV ...]]] [--anytone [RADIO [RADIO ...]]] [--dmrconfig-template [CONF [CONF ...]]]
                   [--farnsworth-template-json [JSON [JSON ...]]] [--gb3gf [RADIO [RADIO ...]]]
                   outdir

dzcb: DMR Zone Channel Builder

positional arguments:
  outdir                Write codeplug files to this directory

optional arguments:
  -h, --help            show this help message and exit
  --pnwdigital          Fetch the latest pnwdigital K7ABD input files
  --seattledmr          Fetch the latest seattledmr K7ABD input files
  --default-k7abd       Include bundled K7ABD input files (simplex + unlicensed)
  --k7abd [DIR [DIR ...]]
                        Specify one or more local directories containing K7ABD CSV files
  --repeaterbook-proximity-csv [CSV [CSV ...]]
                        Fetch repeaters within X distance of POIs defined in a CSV file
  --repeaterbook-state [STATE [STATE ...]]
                        Download repeaters from the given state(s). Default: 'Washington' 'Oregon'
  --repeaterbook-name-format REPEATERBOOK_NAME_FORMAT
                        Python format string used to generate channel names from repeaterbook. See Repeaterbook API response
                        for usable field names. Default: '{Callsign} {Nearest City} {Landmark}'
  --scanlists-json JSON
                        JSON dict mapping scanlist name to list of channel names.
  --include [CSV [CSV ...]]
                        Specify one or more CSV files with object names to include
  --exclude [CSV [CSV ...]]
                        Specify one or more CSV files with object names to exclude
  --order [CSV [CSV ...]]
                        Specify one or more CSV files with object order by name
  --reverse-order [CSV [CSV ...]]
                        Specify one or more CSV files with object order by name (reverse)
  --replacements [CSV [CSV ...]]
                        Specify one or more CSV files with object name replacements
  --anytone [RADIO [RADIO ...]]
                        Generate Anytone CPS CSV files in the 'anytone' subdir for the given radio and CPS versions. If no
                        RADIO+CPS versions are provided, use default set: (578_1_11 868_1_39 878_1_21)
  --dmrconfig-template [CONF [CONF ...]], --dmrconfig [CONF [CONF ...]]
                        Generate dmrconfig conf files in the 'dmrconfig' subdir based on the given template conf files. If no
                        templates are provided, default templates will be used.
  --farnsworth-template-json [JSON [JSON ...]], --editcp [JSON [JSON ...]]
                        Generate Farnsworth editcp JSON format codeplugs in the 'editcp' subdir based on the given template
                        json files. If no templates are provided, default templates will be used.
  --gb3gf [RADIO [RADIO ...]]
                        Generate GB3GF CSV files in the 'gb3gf' subdir for the given radio types. If no radios are provided,
                        use default: (opengd77)
```

## Python API

`dzcb` exposes a python API that is identical to the command line interface
which can be extended to further customize the generation process without
having to directly patch dzcb.

The documentation should improve as the project matures, but for now, see 
[`generate.py`](https://github.com/mycodeplug/example-codeplug/blob/main/input/default/generate.py)
for a real example and [`dzcb.recipe.CodeplugRecipe`](/src/dzcb/recipe.py#L207-L259)
for the implementation details.
        
# Development Usage

## Installation

pip install -e ./dzcb

## Bugs?

Please submit the entire output directory including log files when submitting [issues](/issue/new).


doc/HISTORY.md
# `dzcb` History

`dzcb` is written by Masen Furer KF7HVM <kf7hvm@0x26.net>.

## Motivation

`dzcb` development began in late October 2020 in response to the lack of
existing tools for creating a TYT MD-UV380 codeplug from a matrix of repeater
frequencies and talkgroups.

The [PNWDigital](http://pnwdigital.net) group publishes information on its
repeater system in standard format CSV files (K7ABD format) which are generated
directly from the network cBridges, and I wanted to use these files to directly
generate a cross-platform compatible codeplug targeting the TYT MD-UV380, using
[editcp](https://www.farnsworth.org/dale/codeplug/editcp/).

And thus `dzcb` was born.

## Criticism

> Why did you need to write 3800 lines of code over 6 months to do what I could
> do with an afternoon and a spreadsheet program?

Well, I _personally_ don't like editing codeplugs by hand, neither in a CPS or
in a spreadsheet.

The CPS experience is hit or miss depending on what radio you have and
any work done in a CPS to edit a codeplug will generally not be applicable
to other radio types. Vendor CPS packages are "different enough" from each
other, often buggy, and only support Windows.

The spreadsheet experience can be slightly better (at least you're using find/replace
instead of clicking through lists and buggy forms). However, the problem
with most CSV import formats is _they key on object name_, so renaming objects
can be problematic, and needs to occur in multiple places (and possibly,
multiple separate files).

Finally, neither CPS nor CSV solve the problem of expanding a list of talkgroups
across a single repeater, and the existing tools only support Anytone CPS format.

_(Please email kf7hvm@0x26.net with further criticism and I will respond publicly here)_


## Feature Development

### Repeaterbook

Since [repeaterbook](http://repeaterbook.com) is the primary source of analog
repeater information in the region, I added support for automatically
downloading up-to-date repeater records and formatting those in the same K7ABD
analog format.

Rather than download and assemble the zone/channel CSV by hand (copy/paste from
RB), it made more sense to define a new CSV file with points of interest,
distances, and desired bands to create the channels entries fresh whenever the
codeplug is built.

When repeater information is incorrect in the codeplug, it can be updated on
Repeaterbook to benefit everyone.

### More Output Formats

Along the way, additional output formats were added based on recommendation
or needs.

#### GB3GF OpenGD77

This format is the only good way to represent an OpenGD77 codeplug in CSV format
and is a defacto standard in the opengd77 world.

Because the opengd77 firmware handles static talkgroups much differently than
other common radios, it required a significant restructuring of the internal
representation of the codeplug. Ultimately this architectural change made it
easier to represent a single channel with multiple talkgroups which could later
be expanded into multiple channels.

#### Anytone CSV

Anytone radios are popular and many people prefer to use the official CPS despite
the fact that it only runs on windows. Support for Anytone CPS import format was
added to achieve parity with existing codeplug generators. The implementation is
flexible such that future CPS format changes can be easily accomodated.

#### dmrconfig

On the advice of Dave W7NCX, support for the cross platform cli tool
[`dmrconfig`](https://github.com/OpenRTX/dmrconfig) textual codeplug
format was added.

### Advanced Filtering

In response to user feedback, a new CSV format was created to allow an end user
to customize names, include/exclude objects, and reorder zones, contacts, and
channels within a codeplug.

These modifications are performed against the original source files so there is
no "manual" effort when the upstream data source updates the talkgroup deck,
color code or frequency. The name and ordering will be applied to the updated
channels as before.

### `example-codeplug`

An [example codeplug](https://github.com/mycodeplug/example-codeplug) and
additional documentation was added to facilitate use by others in the
community.


src/dzcb/anytone.py
"""
Write series of CSV files acceptable for import into Anytone CPS tool

Supported CPS versions

    578: 1.11
    868: 1.39
    878: 1.21

"""
import csv
import enum
import logging
from pathlib import Path

from dzcb import AMATEUR_220, COMMERCIAL_UHF, COMMERCIAL_VHF
from dzcb.model import AnalogChannel, Bandwidth, DigitalChannel, uniquify_contacts

logger = logging.getLogger(__name__)

NAME_MAX = 16
SCANLIST_MAX = 50
OFF = "Off"
ON = "On"
NONE = "None"

value_replacements = {
    None: NONE,
    False: OFF,
    True: ON,
}

format_frequency = "{:.5f}".format
format_channel_type = {
    AnalogChannel: "A-Analog",
    DigitalChannel: "D-Digital",
}.get


def format_member_list(members, list_name, expand_members=False):
    d = {list_name: "|".join(m.short_name for m in members)}
    if expand_members:
        d.update(
            {
                list_name
                + " RX Frequency": "|".join(
                    format_frequency(m.frequency) for m in members
                ),
                list_name
                + " TX Frequency": "|".join(
                    format_frequency(m.frequency + m.offset) for m in members
                ),
            }
        )
    return d


def replace_field_names(d, model):
    for orig, repl in model.get("replace_field_names", {}).items():
        if orig in d:
            d[repl] = d[orig]
            del d[orig]
    return d


def remove_fields(d, model):
    for f in model.get("remove_fields", []):
        if f in d:
            del d[f]
    return d


class DMR_MODE(enum.Enum):
    SIMPLEX = 0
    REPEATER = 1
    DUAL_SLOT = 2

    @classmethod
    def value_from(cls, channel):
        if abs(channel.offset) > 0:
            return cls.REPEATER.value
        return cls.SIMPLEX.value


class TXPermit(enum.Enum):
    ALWAYS = "Always"
    CHANNELFREE = "ChannelFree"
    SAMECOLOR = "Same Color Code"
    DIFFERENTCOLOR = "Different Color Code"

    @classmethod
    def value_from(cls, channel):
        if abs(channel.offset) > 0:
            return cls.SAMECOLOR.value
        return cls.ALWAYS.value


# 578/868/878 Common Talkgroups.CSV format
talkgroup_fields = ("No.", "Radio ID", "Name", "Call Type", "Call Alert")
talkgroup_filename = "TalkGroups.CSV"
talkgroup_filename_578_1_11 = "ContactTalkGroups.CSV"

channel_fields_578_1_11 = {
    "No.": None,
    "Channel Name": None,
    "Receive Frequency": None,
    "Transmit Frequency": None,
    "Channel Type": None,
    "Transmit Power": "High",
    "Band Width": "25K",
    "CTCSS/DCS Decode": OFF,
    "CTCSS/DCS Encode": OFF,
    "Contact": "",
    "Contact Call Type": "Group Call",
    "Contact TG/DMR ID": "0",
    "Radio ID": "",
    "Busy Lock/TX Permit": "Always",
    "Squelch Mode": "Carrier",
    "Optional Signal": OFF,
    "DTMF ID": "1",
    "2Tone ID": "1",
    "5Tone ID": "1",
    "PTT ID": OFF,
    "Color Code": "1",
    "Slot": "1",
    "Scan List": NONE,
    "Receive Group List": NONE,
    "PTT Prohibit": OFF,
    "Reverse": OFF,
    "TDMA": OFF,
    "TDMA Adaptive": OFF,
    "AES Digital Encryption": "Normal Encryption",
    "Digital Encryption": OFF,
    "Call Confirmation": OFF,
    "Talk Around(Simplex)": OFF,
    "Work Alone": OFF,
    "Custom CTCSS": "251.1",
    "2TONE Decode": "0",
    "Ranging": OFF,
    "Simplex": OFF,
    "Digi APRS RX": OFF,
    "Analog APRS PTT Mode": OFF,
    "Digital APRS PTT Mode": OFF,
    "APRS Report Type": OFF,
    "Digital APRS Report Channel": "1",
    "Correct Frequency[Hz]": "0",
    "SMS Confirmation": OFF,
    "Exclude Channel From Roaming": "0",
    "DMR MODE": "0",
    "DataACK Disable": "0",
    "R5toneBot": "0",
    "R5ToneEot": "0",
}
channel_fields_868_1_39 = {
    "No.": None,
    "Channel Name": None,
    "Receive Frequency": None,
    "Transmit Frequency": None,
    "Channel Type": None,
    "Transmit Power": "High",
    "Band Width": "25K",
    "CTCSS/DCS Decode": OFF,
    "CTCSS/DCS Encode": OFF,
    "Contact": "",
    "Contact Call Type": "Group Call",
    "Radio ID": "",
    "Busy Lock/TX Permit": "Always",
    "Squelch Mode": "Carrier",
    "Optional Signal": OFF,
    "DTMF ID": "1",
    "2Tone ID": "1",
    "5Tone ID": "1",
    "PTT ID": OFF,
    "Color Code": "1",
    "Slot": "1",
    "CH Scan List": NONE,
    "Receive Group List": NONE,
    "TX Prohibit": OFF,
    "Reverse": OFF,
    "Simplex TDMA": OFF,
    "TDMA Adaptive": OFF,
    "Encryption Type": "Normal Encryption",
    "Digital Encryption": OFF,
    "Call Confirmation": OFF,
    "Talk Around": OFF,
    "Work Alone": OFF,
    "Custom CTCSS": "251.1",
    "2TONE Decode": "0",
    "Ranging": OFF,
    "Through Mode": OFF,
    "APRS Report": OFF,
    "APRS Report Channel": "1",
}
channel_fields_878_1_21 = {
    "No.": None,
    "Channel Name": None,
    "Receive Frequency": None,
    "Transmit Frequency": None,
    "Channel Type": None,
    "Transmit Power": "High",
    "Band Width": "25K",
    "CTCSS/DCS Decode": OFF,
    "CTCSS/DCS Encode": OFF,
    "Contact": "",
    "Contact Call Type": "Group Call",
    "Contact TG/DMR ID": "0",
    "Radio ID": "",
    "Busy Lock/TX Permit": "Always",
    "Squelch Mode": "Carrier",
    "Optional Signal": OFF,
    "DTMF ID": "1",
    "2Tone ID": "1",
    "5Tone ID": "1",
    "PTT ID": OFF,
    "Color Code": "1",
    "Slot": "1",
    "Scan List": NONE,
    "Receive Group List": NONE,
    "PTT Prohibit": OFF,
    "Reverse": OFF,
    "Simplex TDMA": OFF,
    "Slot Suit": OFF,
    "AES Digital Encryption": "Normal Encryption",
    "Digital Encryption": OFF,
    "Call Confirmation": OFF,
    "Talk Around(Simplex)": OFF,
    "Work Alone": OFF,
    "Custom CTCSS": "251.1",
    "2TONE Decode": "0",
    "Ranging": OFF,
    "Through Mode": OFF,
    "Digi APRS RX": OFF,
    "Analog APRS PTT Mode": OFF,
    "Digital APRS PTT Mode": OFF,
    "APRS Report Type": OFF,
    "Digital APRS Report Channel": "1",
    "Correct Frequency[Hz]": "0",
    "SMS Confirmation": OFF,
    "Exclude Channel From Roaming": "0",
    "DMR MODE": "0",
    "DataACK Disable": "0",
    "R5toneBot": "0",
    "R5ToneEot": "0",
}
channel_filename = "Channel.CSV"

scanlist_fields_578_1_11 = {
    "No.": None,
    "Scan List Name": None,
    "Scan Channel Member": None,
    "Scan Channel Member RX Frequency": None,
    "Scan Channel Member TX Frequency": None,
    "Scan Mode": OFF,
    "Priority Channel Select": "Priority Channel Select1",
    "Priority Channel 1": "Current Channel",
    "Priority Channel 1 RX Frequency": "",
    "Priority Channel 1 TX Frequency": "",
    "Priority Channel 2": OFF,
    "Priority Channel 2 RX Frequency": "",
    "Priority Channel 2 TX Frequency": "",
    "Revert Channel": "Selected",
    "Look Back Time A[s]": "2.0",
    "Look Back Time B[s]": "3.0",
    "Dropout Delay Time[s]": "3.1",
    "Dwell Time[s]": "3.1",
}
scanlist_fields_868_1_39 = {
    "No.": None,
    "Scan List Name": None,
    "Scan Channel Member": None,
    "Scan Mode": OFF,
    "Priority Channel Select": "Priority Channel Select1",
    "Priority Channel 1": "Current Channel",
    "Priority Channel 2": OFF,
    "Revert Channel": "Selected",
    "Look Back Time A[s]": "2.0",
    "Look Back Time B[s]": "3.0",
    "Dropout Delay Time[s]": "3.1",
    "Dwell Time[s]": "3.1",
}
scanlist_filename = "ScanList.CSV"

zone_fields_578_1_11 = {
    "No.": None,
    "Zone Name": None,
    "Zone Channel Member": None,
    "Zone Channel Member RX Frequency": None,
    "Zone Channel Member TX Frequency": None,
    "A Channel": None,
    "A Channel RX Frequency": None,
    "A Channel TX Frequency": None,
    "B Channel": None,
    "B Channel RX Frequency": None,
    "B Channel TX Frequency": None,
}
zone_fields_868_1_39 = {
    "No.": None,
    "Zone Name": None,
    "Zone Channel Member": None,
    "A Channel": None,
    "B Channel": None,
}
zone_filename = "Zone.CSV"

SUPPORTED_RADIOS = {
    "578_1_11": dict(
        version="1.11",
        expand_members=True,
        frequency_range=(COMMERCIAL_VHF, AMATEUR_220, COMMERCIAL_UHF),
        channel=channel_fields_578_1_11,
        channel_filename=channel_filename,
        scanlist=scanlist_fields_578_1_11,
        scanlist_filename=scanlist_filename,
        zone=zone_fields_578_1_11,
        zone_filename=zone_filename,
        talkgroup=talkgroup_fields,
        talkgroup_filename=talkgroup_filename_578_1_11,
        replace_field_names={
            "Through Mode": "Simplex",
        },
        remove_fields=[],
    ),
    "868_1_39": dict(
        version="1.39",
        expand_members=False,
        frequency_range=(COMMERCIAL_VHF, COMMERCIAL_UHF),
        channel=channel_fields_868_1_39,
        channel_filename=channel_filename,
        scanlist=scanlist_fields_868_1_39,
        scanlist_filename=scanlist_filename,
        zone=zone_fields_868_1_39,
        zone_filename=zone_filename,
        talkgroup=talkgroup_fields,
        talkgroup_filename=talkgroup_filename,
        replace_field_names={
            "PTT Prohibit": "TX Prohibit",
            "Scan List": "CH Scan List",
        },
        remove_fields=["Contact TG/DMR ID", "DMR MODE"],
    ),
    "878_1_21": dict(
        version="1.21",
        expand_members=True,
        frequency_range=(COMMERCIAL_VHF, COMMERCIAL_UHF),
        channel=channel_fields_878_1_21,
        channel_filename=channel_filename,
        scanlist=scanlist_fields_578_1_11,
        scanlist_filename=scanlist_filename,
        zone=zone_fields_578_1_11,
        zone_filename=zone_filename,
        talkgroup=talkgroup_fields,
        talkgroup_filename=talkgroup_filename,
        replace_field_names={},
        remove_fields=[],
    ),
}

DEFAULT_SUPPORTED_RADIOS = ("578_1_11", "868_1_39", "878_1_21")


def Talkgroup_to_dict(index, talkgroup):
    return {
        "No.": str(index + 1),
        "Radio ID": str(talkgroup.dmrid),
        "Name": talkgroup.name,
        "Call Type": str(talkgroup.kind) + " Call",
        "Call Alert": NONE,
    }


def Talkgroup_to_channel_detail(talkgroup):
    return {
        "Contact": talkgroup.name,
        "Contact Call Type": str(talkgroup.kind) + " Call",
        "Contact TG/DMR ID": talkgroup.dmrid,
        "Slot": str(talkgroup.timeslot),
    }


def AnalogChannel_to_dict(channel):
    return {
        "CTCSS/DCS Decode": channel.tone_decode or OFF,
        "CTCSS/DCS Encode": channel.tone_encode or OFF,
        "Squelch Mode": "CTCSS/DCS" if channel.tone_decode else "Carrier",
        "Busy Lock/TX Permit": OFF,
    }


def DigitalChannel_to_dict(channel):
    d = {
        "Color Code": str(channel.color_code),
        "Busy Lock/TX Permit": TXPermit.value_from(channel),
        "DMR MODE": DMR_MODE.value_from(channel),
        # On the 578 and 878, DMR MODE = "Simplex" (0) channels
        # also have "Simplex=On" and "Through Mode=On" in the
        # exported file. Neither targeted CPS version exposes
        # this setting in the UI. But the 578 1.11 CPS will show
        # DMR MODE "Repeater" unless "Simplex=On"
        # Set it ON for simplex channels, and it will have zero
        # effect because it only makes a difference on split channels
        "Through Mode": OFF if abs(channel.offset) > 0 else ON,
        # TODO: Support group list
    }
    if channel.talkgroup:
        d.update(Talkgroup_to_channel_detail(channel.talkgroup))
    return d


def Channel_to_dict(index, channel, codeplug):
    d = {
        "No.": str(index + 1),
        "Channel Name": channel.short_name,
        "Receive Frequency": format_frequency(channel.frequency),
        "Transmit Frequency": format_frequency(channel.frequency + channel.offset),
        "Channel Type": format_channel_type(type(channel)),
        "Transmit Power": str(channel.power),
        "Band Width": channel.bandwidth.flattened([Bandwidth._25, Bandwidth._125]).value + "K",
        "PTT Prohibit": value_replacements[channel.rx_only],
        "Scan List": channel.scanlist_name(codeplug),
    }
    if isinstance(channel, AnalogChannel):
        d.update(AnalogChannel_to_dict(channel))
    else:
        d.update(DigitalChannel_to_dict(channel))
    return d


def Zone_to_dict(index, zone, expand_members):
    d = {
        "No.": str(index + 1),
        "Zone Name": zone.name,
    }
    d.update(
        format_member_list(
            members=zone.unique_channels,
            list_name="Zone Channel Member",
            expand_members=expand_members,
        )
    )
    for list_name in ("A Channel", "B Channel"):
        d.update(
            format_member_list(
                members=(zone.unique_channels[0],),
                list_name=list_name,
                expand_members=expand_members,
            )
        )
    return d


def ScanList_to_dict(index, scanlist, expand_members):
    d = {
        "No.": str(index + 1),
        "Scan List Name": scanlist.name,
    }
    d.update(
        format_member_list(
            members=scanlist.unique_channels[:SCANLIST_MAX],
            list_name="Scan Channel Member",
            expand_members=expand_members,
        )
    )
    return d


def Codeplug_to_anytone_csv(cp, output_dir, models=None):
    if models is None:
        models = tuple(DEFAULT_SUPPORTED_RADIOS)
    for model_id in models:
        radio_dir = Path(output_dir) / model_id
        radio_dir.mkdir(parents=True, exist_ok=True)
        model = SUPPORTED_RADIOS[model_id]
        # filter down to supported frequency ranges
        mcp = cp.filter(ranges=model["frequency_range"])
        with (radio_dir / model["talkgroup_filename"]).open("w", newline="") as f:
            csvw = csv.DictWriter(f, model["talkgroup"])
            csvw.writeheader()
            for ix, tg in enumerate(
                uniquify_contacts(mcp.contacts, ignore_timeslot=True)
            ):
                csvw.writerow(Talkgroup_to_dict(ix, tg))
        with (radio_dir / model["channel_filename"]).open("w", newline="") as f:
            csvw = csv.DictWriter(f, tuple(model["channel"].keys()))
            csvw.writeheader()
            for ix, channel in enumerate(mcp.channels):
                d = model["channel"].copy()
                d.update(Channel_to_dict(ix, channel, mcp))
                csvw.writerow(replace_field_names(remove_fields(d, model), model))
        with (radio_dir / model["zone_filename"]).open("w", newline="") as f:
            csvw = csv.DictWriter(f, tuple(model["zone"].keys()))
            csvw.writeheader()
            for ix, zone in enumerate(mcp.zones):
                csvw.writerow(Zone_to_dict(ix, zone, model["expand_members"]))
        with (radio_dir / model["scanlist_filename"]).open("w", newline="") as f:
            csvw = csv.DictWriter(f, tuple(model["scanlist"].keys()))
            csvw.writeheader()
            for ix, sl in enumerate(mcp.scanlists):
                d = model["scanlist"].copy()
                d.update(ScanList_to_dict(ix, sl, model["expand_members"]))
                csvw.writerow(d)
        logger.info("Wrote Anytone %s CSV files to '%s'", model_id, radio_dir)
